dts/bindings/pinctrl/quicklogic,eos-s3-pinctrl.yaml

```diff
@@ -0,0 +1,48 @@
+# Copyright (c) 2023 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+  Quicklogic EOS S3 IO MUX binding covers the 46 IOMUX_PAD_x_CTRL registers
+  that can be used to set the direction and the function of a pad.
+  Additionally, it uses func_SEL_REG and func_SEL registers which are needed to
+  configure inputs by writing the pin number to a proper register of the selected function.
+
+  Device pin configuration should be placed in the child nodes of this node.
+  Populate the 'pinmux' field with a triplet consisting of a pin number, its IO
+  function, and register index for that function if it is configured as input,
+  otherwise zero.
+
+  For example, setting pins 44 and 45 for use as UART would look like this:
+
+    #include <dt-bindings/pinctrl/quicklogic-eos-s3-pinctrl.h>
+
+    &pinctrl {
+      uart0_rx_default: uart0_rx_default {
+        pinmux = <QUICKLOGIC_EOS_S3_PINMUX_UART_RX_PAD QUICKLOGIC_EOS_S3_PINMUX_UART_RX_PAD_CFG
+                  QUICKLOGIC_EOS_S3_PINMUX_FUNC_SEL_UART_RX>;
```

Comment:
why do we need 3 macros here? can't a single one be used to specify the UART RX functionality?

Reply:
In order to select an input function, one must first program `IOMUX_PAD_x_CTRL` register,
and then route the signal to functional block using `IOMUX_<function>_SEL`.

The second element in the DTS is used to set proper function to a physical pad.
The third element in the DTS is used for selecting proper `IOMUX_<function_SEL` register, where the pad number must be written.
These values are independent and one cannot be deduced using the other.

Reply:
I mean, can't this all be encoded in 32-bit + dt props? 

By looking at e.g. the RX case we have:

```
#define QUICKLOGIC_EOS_S3_PINMUX_UART_RX_PAD		45

#define QUICKLOGIC_EOS_S3_PINMUX_FUNC_SEL_UART_RX 77

#define QUICKLOGIC_EOS_S3_PINMUX_UART_RX_PAD_CFG \
	(EOS_S3_UART_RXD_SEL_PAD45 | EOS_S3_PAD_OEN_DISABLE \
	 | EOS_S3_PAD_E_4MA | EOS_S3_PAD_REN_ENABLE)
```

PAD 45 is duplicated, for example. And what do all flags like OEN, REN, etc mean?

Reply:
The reason it's done this way is related to the code that uses the HAL and was used for IO MUX:

```
	eos_s3_io_mux(UART_TX_PAD, UART_TX_PAD_CFG);
	eos_s3_io_mux(UART_RX_PAD, UART_RX_PAD_CFG);
	IO_MUX->UART_rxd_SEL = UART_RX_SEL;
```

In order to do the same thing in the driver as depicted in the third line, one must deduce that the input function for `UART_rxd` is being set.
However, for different pads that can serve for the same function, the function values differ:

```
HAL/inc/eoss3_dev.h
1451:#define UART_RXD_SEL_PAD16                          ((uint32_t) (0x00000002))
1452:#define UART_RXD_SEL_PAD45                          ((uint32_t) (0x00000004))
```

Knowing that the function is 0x4 is not enough to conclude that the function of UART_RX is being set here.
Therefore, to ascertain this, we would need to map each (PAD, FUNCTION) pair to a register in `IO_MUX`.
The value assigned would correspond to the shift to this field in the struct:
https://github.com/zephyrproject-rtos/hal_quicklogic/blob/ba7e35f175fcd28199de11f955cb995e1ef20c73/HAL/inc/eoss3_dev.h#L469

The flags used in the code here are utilized to configure `active low output`, `drive strength` and `receive enable` respectively.
The second value provided in the DTS configures the pad as input using these parameters.
While it's a 32-bit value, based on the S3 TRM (pp. 204-205), it appears that only bits [12:0] are actually used. We could pass some information on the more siginifant bits to the driver.

I can see that passing 3 32-bit values seems a bit much as for an IO multiplexer, and I'm sure I can add a mapping of `(PAD, FUNCTION) -> register` (Is this the preferred way?).

Considering the perspecive of both the TRM of EOS S3 and the HAL, it seemed logical to me that the configuration of inputs would require the pad number and two other values. Also, the current solution doesn't add any layer of mapping values between Zephyr and the HAL.

Reply:
Note: Devicetree can’t be influenced by HALs.

Reply:
All right, I've addressed this issue.